generate .dat impurity

sagitta42 opened the pull request:
Rewrote the functions in `impurity.jl` such that first a fit is done using the points as they are in crystal metadata JSONs (corresponding to the Mirion measurement format) - in units of 1e9 e/cm^3 and distance in mm.

Then, if SSD is used, a function is called that would convert the values to e/m^3 and m, and convert the crystal axis (z=0 at seed end) to detector axis (z=0 at p+ contact) using the offset from crystal metadata JSON.
If siggen is used, convert the values to 1e10 e/cm^3 as siggen wants (distance left in mm), and invert the axis, as siggen axis is z=0 at crystal tail (dirtiest); as well as provide correct offset in these coordinates.

Left the option to provide the already presesnt .spe/.dat file and offset in mm to `SiggenSimulator` in simulation settings as alternative to providing crystal metadata path and generating .dat on the fly. Test shows results are consistent (see image, in numbers the depletion voltage differed by 3V due to more precise offset input), but keeping the old option for further tests. Removed the outdated function that created three input parameters inside siggen config for linear profile + quadratic correction (this was not even used by David or Simon for a long time, instead the .dat files were used for a full profile).

One thing: I inserted an error message if both crystal metadata and an already existing spe file are provided, but the code does not stop running after the error. How do I make it stop the simulation?..

![impurity_Mirion_vs_SSD_vs_siggen](https://github.com/legend-exp/LegendGeSim.jl/assets/39589926/2c806b71-25ae-48eb-a06d-307ac1f3d5c7)
![impurity_original_siggen_file-VS-LegendGeSim_generated](https://github.com/legend-exp/LegendGeSim.jl/assets/39589926/60449aca-7690-4746-9d52-821da2f0ce5b)

Commit:
CompatHelper: bump compat for LsqFit to 0.15, (keep existing compat)

Commit:
generate .dat impurity

oschulz commented:
Could we add a test or two for that? Maybe we could add some example files to legend-test-data (modified to remove vendor information) and use these for testing?

fhagemann commented on src/impurity.jl:
```diff
@@ -32,135 +32,205 @@ function SolidStateDetectors.get_impurity_density(
+
+    # get crystal metadata
+    crystal_name = first(meta.name, length(meta.name)-1)
+    crystal_path = joinpath(crystal_metadata_path, crystal_name*".json")
+
+    if !ispath(crystal_path)
+        @error "Crystal $(meta.production.crystal) does not exist in path $crystal_path"
+    end
```
To me, it seems like `crystal_metadata_path` is the path and `crystal_path` is the name of the file?
If that is correct, this should be 
```julia
if !isdir(crystal_metadata_path) 
    @error "Crystal metadata path $crystal_metadata_path does not exist"
end
if !isfile(crystal_path)
    @error "Crystal $(meta.production.crystal) does not exist in path $crystal_metadata_path"
end
```

sagitta42 replied:
The check if crystal metadata folder path exists already happens in `pss.jl` in where the path is read from input settings. At that point the detector name and therefore crystal is not known to the simulator, so I only check the existence of the directory. Then in `imputiry.jl` I check the file itself. This is definitely a bad implementation, also because there's a copy-paste for the check in `SSDSimulator` and `SiggenSimulator`. I will do it the way you wrote - check path and then file in `impurity.jl`, and remove those check from the Simulators.

fhagemann commented on src/pss.jl:
```diff
@@ -107,30 +110,34 @@ Construct SiggeSimulator instance based on simulation
     # check if provided paths exist
     for (param, path) in inputs
-        if (path != "") && !isfile(path)
+        if (path != "") && !(isfile(path) || isdir(path))
```
Is `path` the name of the directory or of the file?

sagitta42 replied:
In case of `:fieldgen_config, :drift_vel, :impurity_profile` it's a file, in case of `:crystal_metadata_path` it's a directory

sagitta42 commented:
@oschulz I guess I could make a "fake" crystal metadata files in examples corresponding to the test ICPC in LegendTestData

oschulz commented:
That would be great.

Commit:
impurities in runtests.jl

sagitta42 commented:
The last commit:

Note: code assumes first impurity measurement is at seed end (z=0), and last impurity measurement at crystal length. While the latter is not important, the first might change the outcome of the fit, as the starting value for the `a` constant is taken as the first measurement.

- made siggen impurity .dat file always have 200 points for uniformity. Points after end of crystal are set to all zero. This is basically how David used to do this also, and is convenient to analyse the files by reading into a 200-point array
- updated `runtests.jl` to include tests of impurity inputs, and tests for both SSD and siggen for field simulation
- update capacitance tolerance to 0.3pF in tests, because with impurity inputs it varies a bit more

Small changes:
- moved checking existence of crystal metadata path from `PSSimulator` classes to `impurity.jl` as suggested by @fhagemann 
- fixed bug in number of points (forgot +1 so the step was not 1mm)
- fixed bug where crystal length was determined as last - first measurement point, to be just last point. So far the first point is always zero from seed end, but just in case, for a good fit crystal length should be truly from seed to tail

sagitta42 commented:
The test cannot find the crystal in the path. The files are now in `legend-testdata`, is this something to do with `LegendTestData.jl`?

fhagemann commented:
Yes. LegendTestData.jl would need to link to the newest commit of `legend-testdata`

fhagemann commented:
I created a PR to update LegendTestData.jl: https://github.com/legend-exp/LegendTestData.jl/pull/5
Once this is merged, the tests should be able to see the crystal metadata files.

fhagemann commented on src/impurity.jl:
```diff
@@ -149,19 +155,25 @@ function impurity_density_model(meta::PropDict, crystal_metadata::PropDict, fit_
     end
+    for i in Npoints+1 : 200
+        imp[i] = 0
+    end
```
Here, you could just define 
```julia
imp = zeros(Float32, 200)
```
and then only run the first loop
```julia
    for i in 1:Npoints
        # divide by 10 to convert from 1e9 e/cm^3 to 1e10 e/cm^3
        # divide by 10 to convert from 1e9 e/cm^3 to 1e10 e/cm^3
        # invert the order of the values because with siggen z=0 is tail (dirtiest part)
        # invert the order of the values because with siggen z=0 is tail (dirtiest part)
        # i.e. z = L in our system (Mirion measurements)
        # i.e. z = L in our system (Mirion measurements)
        imp[Npoints-i+1] = -(a + b * zpoints[i] + c * exp((zpoints[i] - L)/tau)) / 10.
        imp[Npoints-i+1] = -(a + b * zpoints[i] + c * exp((zpoints[i] - L)/tau)) / 10.
    end
```
Then, there would be no need for an additional loop
```julia
    for i in Npoints+1 : 200
        imp[i] = 0
    end
```

sagitta42 replied:
Oh thanks that makes much more sense! Changed it to be that way.

oschulz commented:
New version of LegendTestData.jl is released now.

fhagemann commented:
The tests are re-run. Let's see if they pass now.

sagitta42 commented:
```
[ Info: Cached simulation found. Reading cached fields from cache/B99000A_test_fieldgen_WP.dat
Capacitances: Test Failed at /home/runner/work/LegendGeSim.jl/LegendGeSim.jl/test/runtests.jl:43
  Expression: isapprox(C[1, 2], capacitance, atol = 0.3 * u"pF")
   Evaluated: isapprox(-0.0f0 pF, -7.13 pF; atol = 0.3 pF)
```

So because in `simulation_settings` there is `"cached_name" => "test"`, it read a cached file. Dunno why 0pF and what that file was then. I tested for B99000(A) and got -6.89 pF (thus changed tolerance to 0.3 pF).
How is there even a cached file in tests running online?
Anyway, I would either remove the cached name field from the settings so that it doesn't cache, or add `overwrite=true` in `simulate_fields`. Shall I proceed?

fhagemann commented:
For the tests, we can probably get rid of the caching. It would still be interesting to figure out where this cached file comes from. Do the tests also fail when executed locally (`]test LegendGeSim`)?

fhagemann commented:
Looks like both `B99000A_test_fieldgen_WP.dat` and  `V99000A_test_fieldgen_WP.dat` are empty. 
Therefore, the capacitance is zero.
![image](https://github.com/legend-exp/LegendGeSim.jl/assets/30291312/9e46e0ca-65c2-4da9-8d06-ff0aab12bcea)

fhagemann commented on src/detector.jl:
```diff
@@ -55,9 +55,18 @@ function simulate_detector(det_meta::PropDict, env::Environment, simulator::Sigg
         #...call fieldgen -> will write the wp file
         @info "_|~|_|~|_|~|_ Fieldgen simulation"
-        fieldgen(siggen_config_name; impurity_profile=simulator.impurity_profile, offset_mm=simulator.offset_in_mm)
+        fieldgen(siggen_config_name; impurity_profile=imp_filename, offset_mm=offset)
```
One issue is that `LegendGeSim` is trying to access a function `fieldgen(::AbstractString; impurity_profile, offset_mm)` from `MJDSigGen` that is not part of the latest released version `0.1.0` where it was only `fieldgen(::AbstractString)` without keyword arguments.

I'm getting an error if I try to run this "from scratch" because julia just pulls the latest released version of `MJDSigGen` and not the latest commit on the `master` branch.

oschulz commented:
> How is there even a cached file in tests running online?

Julia CI has the ability to cache certain content, like packages, but I didn't know it could do something like this, automatically.

fhagemann commented on test/runtests.jl:
```diff
@@ -9,47 +9,54 @@ using Unitful
-    )
-    
-
-    pss_table, pss_truth = LegendGeSim.simulate_pulses(det_metadata, path_to_pet_file, environment_settings, simulation_settings) #; n_waveforms=10
```
In order for this to work, the `environment_settings` and `simulation_settings` need to be defined inside the second `@testset`.

sagitta42 commented:
Oh snap, of course I'm a dummy and didn't know how to run locally - should've done that first! If I run locally, they fail in the same manner for siggen with 0 pF Which is even weirder, because this is the first time I do that. 

It doesn't fail without cached name. I did get some error from a bug I introduced while changing the tests in the pss->raw section. Will update both and commit

Commit:
fixed caching bug in tests

sagitta42 commented:
In the last commit: 

Tests:
Removed caching from settings to avoid the strange bug of reading already cached simulation. Now all tests should pass

Small:
- changed `gain_ADC_to_eV` to `gain_ADC_eV` following the current format for variables with divisions e.g. in crystal metadata impurity concentration in 10⁹ e/cm³ saved as `value_in_1e9e_cm3`. `ADC_to_eV` was a bad choice as it seems like it's the opposite, we're converting ADC to eV. Format is discussed also here: https://github.com/legend-exp/legend-data-format-specs/issues/7

- changed `noise_sigma_ADC` to `noise_sigma_in_ADC` (same for `in_keV`) to follow the same format as other parameters `in_unit`

sagitta42 commented:
Error here:
```
  MethodError: no method matching fieldgen(::String; impurity_profile::String, offset_mm::Float32)
  
  Closest candidates are:
    fieldgen(::AbstractString) got unsupported keyword arguments "impurity_profile", "offset_mm"
     @ MJDSigGen ~/.julia/packages/MJDSigGen/KYePY/src/wrappers.jl:120
```

Don't know why, the function exists in `MJDSigGen.jl`:
```julia
function fieldgen(config_filename::AbstractString; impurity_profile="", offset_mm=0)
    setup = read_config(config_filename)

    mkpath(dirname(joinpath(dirname(config_filename), setup.field_name)))
    mkpath(dirname(joinpath(dirname(config_filename), setup.wp_name)))

    if(impurity_profile!="")
        run(`$fieldgen_exe $config_filename -w 1 -p 1 -d 1 -r $impurity_profile -z $offset_mm`)
    else
        run(`$fieldgen_exe $config_filename -w 1 -p 1 -d 1`)
    end
end
```

The tests pass locally.

fhagemann commented:
As said before: because the newest commit of MJDSigGen is not a tagged release.
In the version 0.1.0 pulled in the tests, the function only has the first argument and no keyword arguments.

Either release a new version of MJDSigGen or explicitly pull the master branch for the tests

oschulz commented:
Should I tag a new release of MJDSigGen?

fhagemann commented:
I think this would be useful

oschulz committed:
Update to Documenter v1

oschulz committed:
Update MJDSigGen version bound

Commit:
small fix float32 array

sagitta42 commented:
In the last commit: made the change suggested by @fhagemann with the `zeros()` for the Float32 array

sagitta42 commented:
All checks pass! @oschulz @fhagemann shall we merge?

fhagemann commented on Project.toml:
```diff
@@ -56,8 +56,8 @@ LegendDataManagement = "0.2.1"
-LsqFit = "0.13, 0.14"
-MJDSigGen = "0.1"
+LsqFit = "0.13, 0.14, 0.15"
+MJDSigGen = "0.1, 0.2"
```
Does it make sense to restrict to `MJDSigGen` version `0.2` and drop support for `0.1`?
The changes here might not work with `0.1`.

sagitta42 replied:
That was exactly my question to @oschulz (I think I wrote on Slac): shouldn't it be only 0.2 so that the impurity thing works?

oschulz replied:
Yes, it's fine to require MJDSigGen v0.2

Commit:
MJDSigGen v0.2, impurity errors

sagitta42 commented:
In the latest commit:
1) require MJDSigGen to be v0.2 because as @fhagemann agrees, the siggen impurity file input won't work with v0.1 (that was the whole point v0.2 was tagged)

2) used `throw(ArgumentError("reason"))` as suggested by @fhagemann and @oschulz in `pss.jl` for impurity input checks

sagitta42 committed:
Merge branch 'main' into dev

fhagemann reviewed (approved):
The last merge commit is empty and introduces no changes (?)
But other than this, this should be good to go.

sagitta42 commented:
Thanks @fhagemann! Hmm, the last commit changed MJDSigGen to v0.2 only and chaged `@error` to `throw(...)` (at least I can see those changes)

fhagemann commented:
I was talking about the very last commit (merge main into dev) which seems to do nothing.
But it doesn't matter I guess

fhagemann commented:
Did this PR close #26 ? @sagitta42

sagitta42 commented:
@fhagemann yes
